### PR TITLE
fix: check err before use schedule and duration

### LINKED ISSUE #19605: nil pointer dereference when user provides invalid cron schedule
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`. (caveat below)

**Describe the bug**

The SyncWindow code has no error checking and fails with a nilpointer deref when the user provides an invalid schedule such as `* * * * 2-7`. The invalid code is repeated multiple times in application/v1alpha1/types.go:

```
			schedule, _ := specParser.Parse(w.Schedule)  // returns nil on error
			[...]
			nextWindow := schedule.Next(currentTime.Add(timeZoneOffsetDuration - duration)) // crashes if schedule is nil
```

In one instance they tried to handle the error but don't check sErr until after the nil deref:

```
				schedule, sErr := specParser.Parse(w.Schedule)  // returns nil on error
				[...]
				nextWindow := schedule.Next(currentTime.Add(timeZoneOffsetDuration - duration)) // crashes if schedule is nil

				if !nextWindow.Before(currentTime.Add(timeZoneOffsetDuration)) && sErr == nil && dErr == nil { // never reached if sErr != nil
					inactive = append(inactive, w)
				}
```

You can work around this by copy-pasting all your schedules into the Go Playground to check the format: https://go.dev/play/p/Vfg1RJtIGdA

**To Reproduce**

Create an AppProject with an invalid schedule:

```
apiVersion: argoproj.io/v1alpha1
kind: AppProject
metadata:
  name: default
  namespace: argocd
spec:
  clusterResourceWhitelist:
    - group: '*'
      kind: '*'
  destinations:
    - namespace: '*'
      server: '*'
  sourceRepos:
    - '*'
  syncWindows:
    - kind: allow
      schedule: '0 3 * * 1-5'
      duration: 2h
      timeZone: UTC
      applications:
        - '*'
      manualSync: true
    - kind: deny
      schedule: '* * * * 2-7'
      applications:
      - '*'
      clusters:
      - '*prod*'   # Apply to clusters with "prod" in their name
```

Then try to Sync in the UI.

**Expected behavior**

User sees an error message explaining why their schedule is invalid, eg `end of range (7) above maximum (6): 2-7`.

**Version**

```shell
$ argocd version --port-forward-namespace argocd
FATA[0005] Failed to establish connection to 127.0.0.1:39267: EOF
$ kubectl port-forward svc/argo-cd-argocd-server -n argocd 8080:443 &
Forwarding from 127.0.0.1:8080 -> 8080
Forwarding from [::1]:8080 -> 8080
$ argocd login https://localhost:8080
[0000] dial tcp: address https://localhost:8080: too many colons in address
```

I gave up at this point. I'm using ArgoCD v2.11.5.

**Logs**

```
Recovered from panic: runtime error: invalid memory address or nil pointer dereference
goroutine 213 [running]:
runtime/debug.Stack()n  /usr/local/go/src/runtime/debug/stack.go:24 +0x5e
github.com/argoproj/argo-cd/v2/controller.(*ApplicationController).processAppRefreshQueueItem.func1()
        /go/src/github.com/argoproj/argo-cd/controller/appcontroller.go:1475 +0x54
panic({0x36a5fc0?, 0x7204b20?})
        /usr/local/go/src/runtime/panic.go:920 +0x270
github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*SyncWindows).active(0xc002e3fef0, {0xc003ae5d60?, 0x1e?, 0x744d9e0?})
        /go/src/github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1/types.go:2267 +0x1ce
github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*SyncWindows).Active(0x0?)
        /go/src/github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1/types.go:2249 +0x2b
github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*SyncWindows).CanSync(0xc002e3fef0, 0x0)
        /go/src/github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1/types.go:2423 +0x2c
github.com/argoproj/argo-cd/v2/controller.(*ApplicationController).processAppRefreshQueueItem(0xc000d00e00)
        /go/src/github.com/argoproj/argo-cd/controller/appcontroller.go:1608 +0x1605
github.com/argoproj/argo-cd/v2/controller.(*ApplicationController).Run.func3()
        /go/src/github.com/argoproj/argo-cd/controller/appcontroller.go:824 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
        /go/pkg/mod/k8s.io/apimachinery@v0.26.11/pkg/util/wait/wait.go:157 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0?, {0x50cb3c0, 0xc000f9ff20}, 0x1, 0xc0000cbd40)
        /go/pkg/mod/k8s.io/apimachinery@v0.26.11/pkg/util/wait/wait.go:158 +0xaf
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x0?, 0x0?)
        /go/pkg/mod/k8s.io/apimachinery@v0.26.11/pkg/util/wait/wait.go:135 +0x7f
k8s.io/apimachinery/pkg/util/wait.Until(0x0?, 0x0?, 0x0?)
        /go/pkg/mod/k8s.io/apimachinery@v0.26.11/pkg/util/wait/wait.go:92 +0x1e
created by github.com/argoproj/argo-cd/v2/controller.(*ApplicationController).Run in goroutine 75
        /go/src/github.com/argoproj/argo-cd/controller/appcontroller.go:823 +0x89b
```

See also previous report: https://github.com/argoproj/argo-cd/issues/16701

### PR DESCRIPTION
**Pull request by daengdaengLee**
Fix #19605 

~~I have made some modifications to the related code, though it doesn't fully resolve the issue.~~
I have made some modifications to resolve the issue.
Previously, `schedule` and `duration` were used before checking `sErr` and `dErr`, which could result in a nil pointer exception if an error occurred. (use values in line 2433, check error in line 2435)
I have updated the code to check for errors first and use `continue` if an error is found.

If I've misunderstood something, please let me know in the comments. Thank you! :)

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by blakepettersson** [comment]
Hi @daengdaengLee it looks good so far! Can we just verify that this is actually fixed by adding a test case or two to https://github.com/argoproj/argo-cd/blob/5f23bb60974b3c9e33d0ef0a3626700b6c778cac/pkg/apis/application/v1alpha1/types_test.go#L1932 ?

**Comment by daengdaengLee** [comment]
Hello, @blakepettersson.

I have added unit tests for the incorrect schedule and duration.
Please check it.

Thank you!

File: pkg/apis/application/v1alpha1/types.go [hunk 1]
```diff
@@ -2427,12 +2427,18 @@ func (s *SyncWindows) inactiveAllows(currentTime time.Time) *SyncWindows {
 		for _, w := range *s {
 			if w.Kind == "allow" {
 				schedule, sErr := specParser.Parse(w.Schedule)
+				if sErr != nil {
+					continue
```
Comment (blakepettersson) [inline pkg/apis/application/v1alpha1/types.go:2431]:
```suggestion
                                        log.Warnf("invalid schedule '%v': %s", w.Schedule, err)
					continue
```

File: pkg/apis/application/v1alpha1/types.go [hunk 2]
```diff
@@ -2427,12 +2427,18 @@ func (s *SyncWindows) inactiveAllows(currentTime time.Time) *SyncWindows {
...
+				}
 				duration, dErr := time.ParseDuration(w.Duration)
+				if dErr != nil {
+					continue
```
Comment (blakepettersson) [inline pkg/apis/application/v1alpha1/types.go:2435]:
```suggestion
                                        log.Warnf("invalid schedule '%v': %s", w.Schedule, err)
					continue
```

**Comment by blakepettersson** [comment]
I noticed in the original issue that there are two other places where this is an issue (search for `specParser.Parse` in `types.go`. Could you also handle those ones?

File: pkg/apis/application/v1alpha1/types.go (hunk 1 above)
Comment (daengdaengLee) [reply pkg/apis/application/v1alpha1/types.go:2432]: Thanks for the comment! I've added the log.

File: pkg/apis/application/v1alpha1/types.go (hunk 2 above)
Comment (daengdaengLee) [reply pkg/apis/application/v1alpha1/types.go:2437]: Thanks for the comment! I've added the log.

**Comment by daengdaengLee** [comment]
Hello, @blakepettersson.

Thanks for the comment.
I was planning to submit the other parts in a separate PR.
But if that's your preference, I'll handle everything in this PR!

**Review by blakepettersson — APPROVED** [review]
Looks great, thanks!

**Comment by daengdaengLee** [comment]
Hello, @blakepettersson.

I rebase the PR to the latest master branch.

How can I get a review for the PR?
Please let me know.

Thanks!

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Instead of logging errors in types.go (which is kind of a place for utility functions), could you modify both functions to return an error? That way the consumer can handle the error appropriately, either by logging or by returning the error message to the caller (as in the API Server's `Sync` method).

**Comment by daengdaengLee** [comment]
Hello, @crenshaw-dev!

* return err when sync window is invalid in `types.go`
* keep as close to the original behavior as possible where the functions from `types.go` are used
* add/edit unit tests for functions in `types.go`

Please check and comment if you think differently.
Thanks!

**Review by crenshaw-dev — APPROVED** [review]
lgtm, thank you!

**Comment by crenshaw-dev** [comment]
@daengdaengLee if you'd like to cherry-pick back to 2.10-2.12, looks that those will require manual PRs.

**Pull request merged**